test/extensions/common/wasm/wasm_test.cc

```diff
@@ -1399,6 +1399,91 @@ TEST_P(WasmCommonContextTest, EmptyContext) {
+public:
+  WasmCommonHttpContextTest() = default;
+
+  void setup(const std::string& code, std::string vm_configuration, std::string root_id = "") {
+    setRootId(root_id);
+    setVmConfiguration(vm_configuration);
+    setupBase(std::get<0>(GetParam()), code,
+              [](Wasm* wasm, const std::shared_ptr<Plugin>& plugin) -> ContextBase* {
+                return new NiceMock<TestContext>(wasm, plugin);
+              });
+  }
+  void setupFilter() { setupFilterBase<TestFilter>(); }
+
+  TestContext& rootContext() { return *static_cast<TestContext*>(root_context_); }
+
+  TestFilter& filter() { return *static_cast<TestFilter*>(context_.get()); }
+};
+
+INSTANTIATE_TEST_SUITE_P(Runtimes, WasmCommonHttpContextTest,
+                         Envoy::Extensions::Common::Wasm::runtime_and_cpp_values);
+
+// test that we don't send the local reply twice, even though it's specified in the wasm code
+TEST_P(WasmCommonHttpContextTest, DuplicateLocalReply) {
+  std::string code;
+  if (std::get<0>(GetParam()) != "null") {
+    code = TestEnvironment::readFileToStringForTest(TestEnvironment::substitute(absl::StrCat(
+        "{{ test_rundir }}/test/extensions/common/wasm/test_data/test_context_cpp.wasm")));
+  } else {
+    // The name of the Null VM plugin.
+    code = "CommonWasmTestContextCpp";
+  }
+  EXPECT_FALSE(code.empty());
+
+  setup(code, "", "send local reply twice");
+  setupFilter();
+  EXPECT_CALL(decoder_callbacks_, encodeHeaders_(_, _))
+      .WillOnce([this](Http::ResponseHeaderMap&, bool) { filter().onResponseHeaders(0, false); });
+  EXPECT_CALL(decoder_callbacks_,
+              sendLocalReply(Envoy::Http::Code::OK, testing::Eq("body"), _, _, testing::Eq("ok")));
+
+  // Create in-VM context.
+  filter().onCreate();
+  EXPECT_EQ(proxy_wasm::FilterDataStatus::StopIterationNoBuffer, filter().onRequestBody(0, false));
+}
+
+// test that we don't send the local reply twice when the wasm code panics
+TEST_P(WasmCommonHttpContextTest, LocalReplyWhenPanic) {
+  std::string code;
+  if (std::get<0>(GetParam()) != "null") {
+    code = TestEnvironment::readFileToStringForTest(TestEnvironment::substitute(absl::StrCat(
+        "{{ test_rundir }}/test/extensions/common/wasm/test_data/test_context_cpp.wasm")));
+  } else {
+    // no need test the Null VM plugin.
+    return;
+  }
+  EXPECT_FALSE(code.empty());
+
+  setup(code, "", "panic after sending local reply");
+  setupFilter();
+  // In the case of VM failure, failStream is called, so we need to make sure that we don't send the
+  // local reply twice.
+  EXPECT_CALL(decoder_callbacks_,
+              sendLocalReply(Envoy::Http::Code::ServiceUnavailable, testing::Eq(""), _,
+                             testing::Eq(Grpc::Status::WellKnownGrpcStatus::Unavailable),
+                             testing::Eq("wasm_fail_stream")));
+
+  // Create in-VM context.
+  filter().onCreate();
+  EXPECT_EQ(proxy_wasm::FilterDataStatus::StopIterationNoBuffer, filter().onRequestBody(0, false));
```

Comment:
I think these are redundant for this test

```suggestion
```

Reply:
`filter().onRequestBody` is used to trigger the crash, did you mean like this?
```suggestion
  // Create in-VM context.
  filter().onCreate();
  filter().onRequestBody(0, false);
```

Reply:
oh sorry never mind my comment!